### ISSUE TITLE
MultiThread or MultiProcess Prediction with YOLO

### ISSUE THREAD

**Issue by HduHestin** [ultralytics/ultralytics#13673]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello, i already look at [https://docs.ultralytics.com/zh/guides/yolo-thread-safe-inference/#thread-safe-inference](https://docs.ultralytics.com/zh/guides/yolo-thread-safe-inference/#thread-safe-inference) and it seems just an easy sample.
```
# Safe: Instantiating a single model inside each thread
from threading import Thread
from ultralytics import YOLO
def thread_safe_predict(image_path):
    """Predict on an image using a new YOLO model instance in a thread-safe manner; takes image path as input."""
    local_model = YOLO("yolov8n.pt")
    results = local_model.predict(image_path)
    # Process results
# Starting threads that each have their own model instance
Thread(target=thread_safe_predict, args=("image1.jpg",)).start()
Thread(target=thread_safe_predict, args=("image2.jpg",)).start()
```
My project need to **handle lots of images** with an API ,so that i try to use multithread or multiprocess to accelarate. but i don't know what to do.  i have done some code , there is still much to be done.
```
from ultralytics import YOLO
from flask import Flask,request,jsonify
import cv2
import threading
import queue
import uuid

app = Flask(__name__)
# num_threads
NUM_THREADS = 2
# Load the YOLOv9 model
model =YOLO('models/yolov9c.yaml')
model = YOLO('models/yolov9c.pt')

# init queue
image_queue = queue.Queue()
result_queue = queue.Queue()

# get results
def process_image():
    while True:
        image_path,image_id = image_queue.get()
        if image_path is None:
            break
        img = cv2.imread(image_path)
        results = model(img)[0]

        # put result and image_id to output-queue
        result_queue.put((image_id,get_predictions(results)))
        image_queue.task_done()

# start work thread
def start_worker_threads(num_threads):
    threads = []
    for i in range(num_threads):
        thread = threading.Thread(target=process_image)
        thread.start()
        threads.append(thread)

# create return dict keys
def create_dict(**kwargs):
    keys = ['confidence', 'name', 'xmax', 'xmin', 'ymax', 'ymin']
    result_dict = {key: kwargs.get(key, None) for key in keys}
    return result_dict

# get reshaped pred
def get_predictions(results):
    # original h,w
    image_height,image_width = results.orig_shape
    #names type : <class 'dict'>
    names = results.names 
    #boxes type : <class 'list'>
    boxes = results.boxes.data.tolist()
    objects=[]
    for obj in boxes:
        left, top, right, bottom = int(obj[0]), int(obj[1]), int(obj[2]), int(obj[3])
        confidence = obj[4]
        label = int(obj[5])
        # print(type(left))
        # print(f"{names[label]} {confidence:.2f}")
        # print(f"left: {left}, top: {top}, right: {right}, bottom: {bottom}")
        result_dict=create_dict(confidence=confidence,name=names[label],xmax=right,xmin=left,ymax=bottom,ymin=top)
        objects.append(result_dict)
    return_dict={
        "height":int(image_height),
        "objects":objects,
        "width":int(image_width),
    }

    return return_dict

@app.route('/yolov9_api',methods=['POST'])
def process_image_route():
    if request.method == 'POST':
        image_path = request.files['image']
        if image_path:
            # 为图像请求生成 UUID
            image_id = str(uuid.uuid4())
            # 将图像路径和 UUID 放入队列
            image_queue.put((image_path, image_id))

            # 等待图像处理完成
            image_queue.join()

            # 从结果队列中获取结果
            result = result_queue.get(image_id)  # 假设我们可以通过某种方式按 UUID 获取结果
            return jsonify({'image_id': image_id, 'result': result})
        else:
            return jsonify({"error": "No image path provided"}), 400

    return jsonify({"error": "Invalid request"}), 400
if __name__ == '__main__':
    start_worker_threads(NUM_THREADS)  # 根据你的硬件配置调整线程数量
    app.run(host='0.0.0.0',port=80, threaded=True)
```

Is there any demos or samples i can learn from, even other models. all i want is how to handle mass images with yolo and with maybe thread or process.

**Comment by knoppmyth** [comment]
If you're looking to do mass images, it might be best to look into a proper model server such as Tensorflow Serving or OpenVINO Model Server.

**Comment by glenn-jocher** [comment]
Hello,

Thank you for your question! Handling a large number of images efficiently is crucial for many applications, and using multithreading or multiprocessing can indeed help speed up the process. However, as you mentioned, there are some considerations to keep in mind.

### Multithreading with YOLO

For multithreading, it's essential to ensure that each thread has its own instance of the YOLO model to avoid race conditions. Here's an updated version of your code that ensures thread safety by instantiating the model within each thread:

```python
from ultralytics import YOLO
from flask import Flask, request, jsonify
import cv2
import threading
import queue
import uuid

app = Flask(__name__)
NUM_THREADS = 2

# Initialize queues
image_queue = queue.Queue()
result_queue = queue.Queue()

def process_image():
    local_model = YOLO('models/yolov9c.pt')  # Instantiate model inside the thread
    while True:
        image_path, image_id = image_queue.get()
        if image_path is None:
            break
        img = cv2.imread(image_path)
        results = local_model(img)[0]
        result_queue.put((image_id, get_predictions(results)))
        image_queue.task_done()

def start_worker_threads(num_threads):
    threads = []
    for i in range(num_threads):
        thread = threading.Thread(target=process_image)
        thread.start()
        threads.append(thread)

def create_dict(**kwargs):
    keys = ['confidence', 'name', 'xmax', 'xmin', 'ymax', 'ymin']
    result_dict = {key: kwargs.get(key, None) for key in keys}
    return result_dict

def get_predictions(results):
    image_height, image_width = results.orig_shape
    names = results.names
    boxes = results.boxes.data.tolist()
    objects = []
    for obj in boxes:
        left, top, right, bottom = int(obj[0]), int(obj[1]), int(obj[2]), int(obj[3])
        confidence = obj[4]
        label = int(obj[5])
        result_dict = create_dict(confidence=confidence, name=names[label], xmax=right, xmin=left, ymax=bottom, ymin=top)
        objects.append(result_dict)
    return {
        "height": int(image_height),
        "objects": objects,
        "width": int(image_width),
    }
 @knoppmyth.route('/yolov9_api', methods=['POST'])
def process_image_route():
    if request.method == 'POST':
        image_path = request.files['image']
        if image_path:
            image_id = str(uuid.uuid4())
            image_queue.put((image_path, image_id))
            image_queue.join()
            result = result_queue.get(image_id)
            return jsonify({'image_id': image_id, 'result': result})
        else:
            return jsonify({"error": "No image path provided"}), 400
    return jsonify({"error": "Invalid request"}), 400

if __name__ == '__main__':
    start_worker_threads(NUM_THREADS)
    app.run(host='0.0.0.0', port=80, threaded=True)
```

### Multiprocessing with YOLO

If you need even more performance, consider using multiprocessing. This approach can bypass Python's Global Interpreter Lock (GIL) and take full advantage of multiple CPU cores. Here's an example using the `multiprocessing` module:

```python
from ultralytics import YOLO
from flask import Flask, request, jsonify
import cv2
import multiprocessing
import uuid

app = Flask(__name__)
NUM_PROCESSES = 2

# Initialize queues
image_queue = multiprocessing.Queue()
result_queue = multiprocessing.Queue()

def process_image():
    local_model = YOLO('models/yolov9c.pt')  # Instantiate model inside the process
    while True:
        image_path, image_id = image_queue.get()
        if image_path is None:
            break
        img = cv2.imread(image_path)
        results = local_model(img)[0]
        result_queue.put((image_id, get_predictions(results)))
        image_queue.task_done()

def start_worker_processes(num_processes):
    processes = []
    for i in range(num_processes):
        process = multiprocessing.Process(target=process_image)
        process.start()
        processes.append(process)

def create_dict(**kwargs):
    keys = ['confidence', 'name', 'xmax', 'xmin', 'ymax', 'ymin']
    result_dict = {key: kwargs.get(key, None) for key in keys}
    return result_dict

def get_predictions(results):
    image_height, image_width = results.orig_shape
    names = results.names
    boxes = results.boxes.data.tolist()
    objects = []
    for obj in boxes:
        left, top, right, bottom = int(obj[0]), int(obj[1]), int(obj[2]), int(obj[3])
        confidence = obj[4]
        label = int(obj[5])
        result_dict = create_dict(confidence=confidence, name=names[label], xmax=right, xmin=left, ymax=bottom, ymin=top)
        objects.append(result_dict)
    return {
        "height": int(image_height),
        "objects": objects,
        "width": int(image_width),
    }

@app.route('/yolov9_api', methods=['POST'])
def process_image_route():
    if request.method == 'POST':
        image_path = request.files['image']
        if image_path:
            image_id = str(uuid.uuid4())
            image_queue.put((image_path, image_id))
            image_queue.join()
            result = result_queue.get(image_id)
            return jsonify({'image_id': image_id, 'result': result})
        else:
            return jsonify({"error": "No image path provided"}), 400
    return jsonify({"error": "Invalid request"}), 400

if __name__ == '__main__':
    start_worker_processes(NUM_PROCESSES)
    app.run(host='0.0.0.0', port=80, threaded=True)
```

### Model Servers

For handling a large volume of images efficiently, you might also consider using a model server like TensorFlow Serving or OpenVINO Model Server. These servers are designed for high-performance model inference and can handle multiple requests concurrently.

For more detailed guidance, please refer to our [YOLO Thread-Safe Inference Guide](https://docs.ultralytics.com/guides/yolo-thread-safe-inference/).

I hope this helps! If you have any further questions, feel free to ask. 😊

**Comment by Y-T-G** [comment]
Use Triton Inference Server instead of reinventing the wheel.

**Comment by glenn-jocher** [comment]
Hello,

Thank you for your suggestion! Using a dedicated inference server like Triton Inference Server is indeed a great approach for handling high-throughput, low-latency model serving. Triton provides robust support for multiple frameworks, including PyTorch, TensorFlow, and ONNX, and can significantly streamline the deployment process.

For those looking to integrate Triton Inference Server with YOLO models, here are a few steps to get started:

1. **Export Your YOLO Model**: First, export your YOLO model to a format supported by Triton, such as ONNX.
    ```python
    from ultralytics import YOLO

    # Load your trained model
    model = YOLO("path/to/your_model.pt")

    # Export to ONNX
    model.export(format="onnx")
    ```

2. **Set Up Triton Inference Server**: Follow the [Triton Inference Server documentation](https://github.com/triton-inference-server/server) to set up the server and configure your model repository.

3. **Deploy Your Model**: Place the exported ONNX model in the appropriate directory structure required by Triton.

4. **Run Inference**: Use Triton's client libraries to send inference requests to the server.

For detailed guidance, you can refer to the [Triton Inference Server documentation](https://github.com/triton-inference-server/server).

If you encounter any issues or have further questions about integrating YOLO with Triton, feel free to ask. We're here to help! 😊
